File: backend/learning/Learning/Infrastructure/Messaging/Events/BaseEvent.cs
namespace Learning.Infrastructure.Messaging.Events;

public interface IBaseEvent
{
    public string EventId { get; init; }
    public string EventType { get; init; }
    public string CorrelationId { get; init; }
    public string Timestamp { get; init; }
}

public abstract record BaseEvent : IBaseEvent
{
    public string EventId { get; init; } = Guid.NewGuid().ToString();
    public string Timestamp { get; init; } = DateTime.UtcNow.ToString("o");
    public string CorrelationId { get; init; } = Guid.NewGuid().ToString();
    public abstract string EventType { get; init; }
}

File: backend/auth/src/main/java/com/example/auth/infrastructure/rabbitmq/events/BaseEvent.java
package com.example.auth.infrastructure.rabbitmq.events;

import java.io.Serializable;
import java.time.Instant;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Getter
@ToString
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.PUBLIC_ONLY)
@JsonIgnoreProperties(ignoreUnknown = true)
public abstract class BaseEvent implements Serializable {
    private String eventType;
    private String eventId;
    @Setter
    private String correlationId;
    private String timestamp;

    protected BaseEvent() {
        this.eventId = UUID.randomUUID().toString();
        this.correlationId = UUID.randomUUID().toString();
        this.timestamp = Instant.now().toString();
    }

    protected void setEventType(String eventType) {
        this.eventType = eventType;
    }
}